ql/src/java/org/apache/hadoop/hive/ql/hooks/ATSHook.java

```diff
@@ -0,0 +1,492 @@
+
+  private static final Logger LOG = LoggerFactory.getLogger(ATSHook.class.getName());
+  private boolean isATSEnabled = false;
+  private static final Object LOCK = new Object();
+  private static final int VERSION = 2;
+  private static ExecutorService executor;
+  private static ExecutorService senderExecutor;
+  private static TimelineClient timelineClient;
+  private enum EntityTypes { HIVE_QUERY_ID };
+  private enum EventTypes { QUERY_SUBMITTED, QUERY_COMPLETED };
+  private static final String ATS_DOMAIN_PREFIX = "hive_";
+  private static boolean defaultATSDomainCreated = false;
+  private static final String DEFAULT_ATS_DOMAIN = "hive_default_ats";
+
+  @VisibleForTesting
+  enum OtherInfoTypes {
+    QUERY, STATUS, TEZ, MAPRED, INVOKER_INFO, SESSION_ID, THREAD_NAME, VERSION,
+    CLIENT_IP_ADDRESS, HIVE_ADDRESS, HIVE_INSTANCE_TYPE, CONF, PERF, LLAP_APP_ID
+  };
+  private enum ExecutionMode {
+    MR, TEZ, LLAP, SPARK, NONE
+  };
+  private enum PrimaryFilterTypes {
+    user, requestuser, operationid, executionmode, tablesread, tableswritten, queue
+  };
+
+  private static final int WAIT_TIME = 3;
+
+  private static final String[] PERF_KEYS = new String[] {
+    PerfLogger.PARSE, PerfLogger.COMPILE, PerfLogger.ANALYZE, PerfLogger.OPTIMIZER,
+    PerfLogger.GET_SPLITS, PerfLogger.RUN_TASKS,
+  };
+
+  private static void setupAtsExecutor(HiveConf conf) {
+    synchronized(LOCK) {
+      if (executor == null) {
+
+        // The call to ATS appears to block indefinitely, blocking the ATS thread while
+        // the hook continues to submit work to the ExecutorService with each query.
+        // Over time the queued items can cause OOM as the HookContext seems to contain
+        // some items which use a lot of memory.
+        // Prevent this situation by creating executor with bounded capacity -
+        // the event will not be sent to ATS if there are too many outstanding work submissions.
+        int queueCapacity = conf.getIntVar(ConfVars.ATS_HOOK_QUEUE_CAPACITY);
+
+        // Executor to create the ATS events.
+        // This can use significant resources and should not be done on the main query thread.
+        LOG.info("Creating ATS executor queue with capacity " + queueCapacity);
+        BlockingQueue<Runnable> queue = new LinkedBlockingQueue<Runnable>(queueCapacity);
+        ThreadFactory threadFactory = new ThreadFactoryBuilder().setDaemon(true).setNameFormat("ATS Logger %d").build();
+        executor = new ThreadPoolExecutor(1, 1, 0, TimeUnit.MILLISECONDS, queue, threadFactory);
+
+        // Create a separate thread to send the events.
+        // Keep separate from the creating events in case the send blocks.
+        BlockingQueue<Runnable> senderQueue = new LinkedBlockingQueue<Runnable>(queueCapacity);
+        senderExecutor = new ThreadPoolExecutor(1, 1, 0, TimeUnit.MILLISECONDS, senderQueue, threadFactory);
+
+        YarnConfiguration yarnConf = new YarnConfiguration();
```

Comment:
Is that only supported for YARN deployments?

Reply:
I'm not sure why this hook was dropped, but I think it is useful. 
But my question was more about if we could tweak the YARN stuff to be cloud-native.